# Esercizio 3 - Soluzione

### 1) Parametri in input

### 2) Importazione del modulo `re`

### 3) Definizione della funzione `format_fasta()`

La funzione prende come argomenti un *header* `FASTA` (contenente un simbolo `>` all'inizio), una sequenza nucleotidica (o di proteina) e la lunghezza dei *record* della sequenza. La funzione restituisce una stringa contenente la sequenza in formato `FASTA` separata in *record* di un numero di caratteri pari al valore del terzo argomento.

*Suggerimento*: usare la funzione `compile()` per produrre un'espressione regolare variabile.

### 4) Lettura del file `EMBL` in un'unica stringa

*Suggerimento*: leggere il file in un'unica stringa usando il metodo `read()`.

In [5]:
print(file_str)

ID   M10051; SV 1; linear; mRNA; STD; HUM; 4723 BP.
XX
AC   M10051;
XX
DT   02-JUL-1986 (Rel. 09, Created)
DT   14-NOV-2006 (Rel. 89, Last updated, Version 7)
XX
DE   Human insulin receptor mRNA, complete cds.
XX
KW   insulin receptor; tyrosine kinase.
XX
OS   Homo sapiens (human)
OC   Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;
OC   Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;
OC   Homo.
XX
RN   [1]
RP   1-4723
RX   DOI; 10.1016/0092-8674(85)90334-4.
RX   PUBMED; 2859121.
RA   Ebina Y., Ellis L., Jarnagin K., Edery M., Graf L., Clauser E., Ou J.-H.,
RA   Masiarz F., Kan Y.W., Goldfine I.D., Roth R.A., Rutter W.J.;
RT   "The human insulin receptor cDNA: the structural basis for
RT   hormone-activated transmembrane signalling";
RL   Cell 40(4):747-758(1985).
XX
DR   MD5; e4e6ebf2e723a500c1dd62385c279351.
DR   Ensembl-Gn; ENSG00000171105; homo_sapiens.
DR   Ensembl-Tr; ENST00000302850; homo_sapiens.
DR   Ensembl-Tr; ENST0000034

### 5) Estrazione dell'identificatore univoco e dell'organismo relativo all'entry.

Estrarre dal *record* `ID`:

    ID   M10051; SV 1; linear; mRNA; STD; HUM; 4723 BP.
    
l'identificatore univoco e l'organismo e assegnarli alle variabili `identifier` e `organism`.

a) Estrarre l'identificatore.

In [47]:
identifier

'M10051'

b) Estrarre l'organismo.

In [51]:
organism

'HUM'

c) ALTERNATIVA: Estrarre l'identificatore e l'organismo con un'unica RE.

    ID   M10051; SV 1; linear; mRNA; STD; HUM; 4723 BP.


In [53]:
identifier

'M10051'

In [54]:
organism

'HUM'

### 6) Estrazione della sequenza nucleotidica

a) Costruire la lista dei *record* della sequenza nucleotidica, escludendo da ognuno di essi gli spazi iniziali, gli spazi finali e l'intero finale.

    ggggggctgc gcggccgggt cggtgcgcac acgagaagga cgcgcggccc ccagcgctct        60

In [58]:
seq_row_list

['ggggggctgc gcggccgggt cggtgcgcac acgagaagga cgcgcggccc ccagcgctct',
 'tgggggccgc ctcggagcat gacccccgcg ggccagcgcc gcgcgcctga tccgaggaga',
 'ccccgcgctc ccgcagccat gggcaccggg ggccggcggg gggcggcggc cgcgccgctg',
 'ctggtggcgg tggccgcgct gctactgggc gccgcgggcc acctgtaccc cggagaggtg',
 'tgtcccggca tggatatccg gaacaacctc actaggttgc atgagctgga gaattgctct',
 'gtcatcgaag gacacttgca gatactcttg atgttcaaaa cgaggcccga agatttccga',
 'gacctcagtt tccccaaact catcatgatc actgattact tgctgctctt ccgggtctat',
 'gggctcgaga gcctgaagga cctgttcccc aacctcacgg tcatccgggg atcacgactg',
 'ttctttaact acgcgctggt catcttcgag atggttcacc tcaaggaact cggcctctac',
 'aacctgatga acatcacccg gggttctgtc cgcatcgaga agaacaatga gctctgttac',
 'ttggccacta tcgactggtc ccgtatcctg gattccgtgg aggataatca catcgtgttg',
 'aacaaagatg acaacgagga gtgtggagac atctgtccgg gtaccgcgaa gggcaagacc',
 'aactgccccg ccaccgtcat caacgggcag tttgtcgaac gatgttggac tcatagtcac',
 'tgccagaaag tttgcccgac catctgtaag tcacacggct gcaccgccga aggcctctgt',
 'tgccacagcg agtgcct

b) Concatenare i singoli pezzi per ottenere la sequenza nucleotidica in unica stringa (in lettere maiuscole).

*Suggerimento*: il metodo `join()` degli oggetti `str` restituisce la concatenazione delle stringhe presenti nella lista passata come argomento usando la stringa invocante come separatore.

In [16]:
nucleotide_sequence

'GGGGGGCTGCGCGGCCGGGTCGGTGCGCACACGAGAAGGACGCGCGGCCCCCAGCGCTCTTGGGGGCCGCCTCGGAGCATGACCCCCGCGGGCCAGCGCCGCGCGCCTGATCCGAGGAGACCCCGCGCTCCCGCAGCCATGGGCACCGGGGGCCGGCGGGGGGCGGCGGCCGCGCCGCTGCTGGTGGCGGTGGCCGCGCTGCTACTGGGCGCCGCGGGCCACCTGTACCCCGGAGAGGTGTGTCCCGGCATGGATATCCGGAACAACCTCACTAGGTTGCATGAGCTGGAGAATTGCTCTGTCATCGAAGGACACTTGCAGATACTCTTGATGTTCAAAACGAGGCCCGAAGATTTCCGAGACCTCAGTTTCCCCAAACTCATCATGATCACTGATTACTTGCTGCTCTTCCGGGTCTATGGGCTCGAGAGCCTGAAGGACCTGTTCCCCAACCTCACGGTCATCCGGGGATCACGACTGTTCTTTAACTACGCGCTGGTCATCTTCGAGATGGTTCACCTCAAGGAACTCGGCCTCTACAACCTGATGAACATCACCCGGGGTTCTGTCCGCATCGAGAAGAACAATGAGCTCTGTTACTTGGCCACTATCGACTGGTCCCGTATCCTGGATTCCGTGGAGGATAATCACATCGTGTTGAACAAAGATGACAACGAGGAGTGTGGAGACATCTGTCCGGGTACCGCGAAGGGCAAGACCAACTGCCCCGCCACCGTCATCAACGGGCAGTTTGTCGAACGATGTTGGACTCATAGTCACTGCCAGAAAGTTTGCCCGACCATCTGTAAGTCACACGGCTGCACCGCCGAAGGCCTCTGTTGCCACAGCGAGTGCCTGGGCAACTGTTCTCAGCCCGACGACCCCACCAAGTGCGTGGCCTGCCGCAACTTCTACCTGGACGGCAGGTGTGTGGAGACCTGCCCGCCCCCGTACTACCACTTCCAGGACTGGCGCTGTGTGAACTTCAGCTTCTGCCAG

### 7) Verifica della composizione della sequenza nucleotidica rispetto a quella dichiarata nell'*entry* `EMBL`

a) Estrarre dal *record* `SQ`:

    SQ   Sequence 4723 BP; 1068 A; 1298 C; 1311 G; 1046 T; 0 other;
    
la composizione della sequenza nucleotidica in un dizionario:

- *chiave*: nucleotide
- *valore*: numero di occorrenze della chiave

Estrarre il record `SQ`.

In [18]:
sq_record

'SQ   Sequence 4723 BP; 1068 A; 1298 C; 1311 G; 1046 T; 0 other;'

Creare il dizionario.

In [26]:
entry_seq_composition

{'A': 1068, 'C': 1298, 'G': 1311, 'T': 1046}

b) Estrarre un dizionario analogo dalla sequenza nucleotidica.

In [28]:
seq_composition

{'G': 1311, 'C': 1298, 'T': 1046, 'A': 1068}

c) Confrontare i due dizionari

### 8) Creazione della sequenza nucleotidica in formato `FASTA`

Produrre la sequenza nucleotidica in formato `FASTA` con il seguente *header*:

    >M10051-HUM

a) Sostituire ogni simbolo `T` con un simbolo `U`.

In [31]:
nucleotide_sequence

'GGGGGGCUGCGCGGCCGGGUCGGUGCGCACACGAGAAGGACGCGCGGCCCCCAGCGCUCUUGGGGGCCGCCUCGGAGCAUGACCCCCGCGGGCCAGCGCCGCGCGCCUGAUCCGAGGAGACCCCGCGCUCCCGCAGCCAUGGGCACCGGGGGCCGGCGGGGGGCGGCGGCCGCGCCGCUGCUGGUGGCGGUGGCCGCGCUGCUACUGGGCGCCGCGGGCCACCUGUACCCCGGAGAGGUGUGUCCCGGCAUGGAUAUCCGGAACAACCUCACUAGGUUGCAUGAGCUGGAGAAUUGCUCUGUCAUCGAAGGACACUUGCAGAUACUCUUGAUGUUCAAAACGAGGCCCGAAGAUUUCCGAGACCUCAGUUUCCCCAAACUCAUCAUGAUCACUGAUUACUUGCUGCUCUUCCGGGUCUAUGGGCUCGAGAGCCUGAAGGACCUGUUCCCCAACCUCACGGUCAUCCGGGGAUCACGACUGUUCUUUAACUACGCGCUGGUCAUCUUCGAGAUGGUUCACCUCAAGGAACUCGGCCUCUACAACCUGAUGAACAUCACCCGGGGUUCUGUCCGCAUCGAGAAGAACAAUGAGCUCUGUUACUUGGCCACUAUCGACUGGUCCCGUAUCCUGGAUUCCGUGGAGGAUAAUCACAUCGUGUUGAACAAAGAUGACAACGAGGAGUGUGGAGACAUCUGUCCGGGUACCGCGAAGGGCAAGACCAACUGCCCCGCCACCGUCAUCAACGGGCAGUUUGUCGAACGAUGUUGGACUCAUAGUCACUGCCAGAAAGUUUGCCCGACCAUCUGUAAGUCACACGGCUGCACCGCCGAAGGCCUCUGUUGCCACAGCGAGUGCCUGGGCAACUGUUCUCAGCCCGACGACCCCACCAAGUGCGUGGCCUGCCGCAACUUCUACCUGGACGGCAGGUGUGUGGAGACCUGCCCGCCCCCGUACUACCACUUCCAGGACUGGCGCUGUGUGAACUUCAGCUUCUGCCAG

b) Creare l'*header*.

b) Creare la sequenza in formato `FASTA`.

In [34]:
print(nucleotide_fasta_sequence)

>M10051-HUM
GGGGGGCUGCGCGGCCGGGUCGGUGCGCACACGAGAAGGACGCGCGGCCCCCAGCGCUCUUGGGGGCCGCCUCGGAGCAU
GACCCCCGCGGGCCAGCGCCGCGCGCCUGAUCCGAGGAGACCCCGCGCUCCCGCAGCCAUGGGCACCGGGGGCCGGCGGG
GGGCGGCGGCCGCGCCGCUGCUGGUGGCGGUGGCCGCGCUGCUACUGGGCGCCGCGGGCCACCUGUACCCCGGAGAGGUG
UGUCCCGGCAUGGAUAUCCGGAACAACCUCACUAGGUUGCAUGAGCUGGAGAAUUGCUCUGUCAUCGAAGGACACUUGCA
GAUACUCUUGAUGUUCAAAACGAGGCCCGAAGAUUUCCGAGACCUCAGUUUCCCCAAACUCAUCAUGAUCACUGAUUACU
UGCUGCUCUUCCGGGUCUAUGGGCUCGAGAGCCUGAAGGACCUGUUCCCCAACCUCACGGUCAUCCGGGGAUCACGACUG
UUCUUUAACUACGCGCUGGUCAUCUUCGAGAUGGUUCACCUCAAGGAACUCGGCCUCUACAACCUGAUGAACAUCACCCG
GGGUUCUGUCCGCAUCGAGAAGAACAAUGAGCUCUGUUACUUGGCCACUAUCGACUGGUCCCGUAUCCUGGAUUCCGUGG
AGGAUAAUCACAUCGUGUUGAACAAAGAUGACAACGAGGAGUGUGGAGACAUCUGUCCGGGUACCGCGAAGGGCAAGACC
AACUGCCCCGCCACCGUCAUCAACGGGCAGUUUGUCGAACGAUGUUGGACUCAUAGUCACUGCCAGAAAGUUUGCCCGAC
CAUCUGUAAGUCACACGGCUGCACCGCCGAAGGCCUCUGUUGCCACAGCGAGUGCCUGGGCAACUGUUCUCAGCCCGACG
ACCCCACCAAGUGCGUGGCCUGCCGCAACUUCUACCUGGACGGCAGGUGUGUGGAGACCUGCCCGCCCCCGUACUACCAC
UUCCAGGACUGGCGCU

Stampare la sequenza su file.

### 9) Estrazione della sequenza della proteina

a) Estrarre il prefisso della proteina contenuto nel *record*:

    FT                   /translation="MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLT

In [37]:
protein_prefix

'MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLT'

b) Costruire la lista di tutti gli altri record della proteina (compreso l'ultimo):

    FT                   RLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFP
    
**Attenzione all'ultimo**:

    FT                   DGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSNPS"

che termina con doppi apici `"`.

In [39]:
protein_list

['RLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFP',
 'NLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRI',
 'LDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTI',
 'CKSHGCTAEGLCCHSECLGNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCV',
 'NFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCH',
 'LLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRR',
 'SYALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTTTQGKLFFHYN',
 'PKLCLSEIHKMEEVSGTKGRQERNDIALKTNGDKASCENELLKFSYIRTSFDKILLRWE',
 'PYWPPDFRDLLGFMLFYKEAPYQNVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHP',
 'GWLMRGLKPWTQYAIFVKTLVTFSDERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSS',
 'SQIILKWKPPSDPNGNITHYLVFWERQAEDSELFELDYCLKGLKLPSRTWSPPFESEDS',
 'QKHNQSEYEDSAGECCSCPKTDSQILKELEESSFRKTFEDYLHNVVFVPRKTSSGTGAE',
 'DPRPSRKRRSLGDVGNVTVAVPTVAAFPNTSSTSVPTSPEEHRPFEKVVNKESLVISGL',
 'RHFTGYRIELQACNQDTPEERCSVAAYVSARTMPEAKADDIVGPVTHEIFENNVVHLMW',
 'QEPKEPNGLIVLYEVSYRRYGDEELHLCVSRKHFALERGCRLRGLSPGNYSVRIRATSL',
 'AGNGSWTEPTYFYVTDYLDVPSNIAKIIIGPLIFVFLF

b) Aggiungere in testa alla lista il prefisso trovato prima e concatenare tutti gli elementi della lista per ottenere la sequenza della proteina in un'unica stringa.

In [41]:
protein_sequence

'MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLTRLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRILDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECLGNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRSYALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTTTQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQERNDIALKTNGDKASCENELLKFSYIRTSFDKILLRWEPYWPPDFRDLLGFMLFYKEAPYQNVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHPGWLMRGLKPWTQYAIFVKTLVTFSDERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSSSQIILKWKPPSDPNGNITHYLVFWERQAEDSELFELDYCLKGLKLPSRTWSPPFESEDSQKHNQSEYEDSAGECCSCPKTDSQILKELEESSFRKTFEDYLHNVVFVPRKTSSGTGAEDPRPSRKRRSLGDVGNVTVAVPTVAAFPNTSSTSVPTSPEEHRPFEKVVNKESLVISGLRHFTGYRIELQACNQDTPEERCSVAAYVSARTMPEAKADDIVGPVTHEIFENNVVHLMWQEPKEPNGLIVLYEVSYRRYGDEELHLCVSRKHFALERGCRLRGLSPGNYSVRIRATSLAGNGSWTEPTYFYVTDYLDVPSNIAKIIIGPLIFVFLFSVVIGSIYLFLRKRQPDGPLGPLYASSNPEY

### 10) Creazione della sequenza della proteina in formato `FASTA`

Produrre la sequenza della proteina in formato `FASTA` con il seguente *header*:

    >M10051-HUM; len = 1382

a) Creare l'*header*.

b) Creare la sequenza della proteina in formato `FASTA`.

In [44]:
print(protein_fasta_sequence)

>M10051-HUM; len = 1382
MGTGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLTRLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPK
LIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDW
SRILDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECL
GNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYT
MNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRS
YALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTTTQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQE
RNDIALKTNGDKASCENELLKFSYIRTSFDKILLRWEPYWPPDFRDLLGFMLFYKEAPYQNVTEFDGQDACGSNSWTVVD
IDPPLRSNDPKSQNHPGWLMRGLKPWTQYAIFVKTLVTFSDERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSSSQIIL
KWKPPSDPNGNITHYLVFWERQAEDSELFELDYCLKGLKLPSRTWSPPFESEDSQKHNQSEYEDSAGECCSCPKTDSQIL
KELEESSFRKTFEDYLHNVVFVPRKTSSGTGAEDPRPSRKRRSLGDVGNVTVAVPTVAAFPNTSSTSVPTSPEEHRPFEK
VVNKESLVISGLRHFTGYRIELQACNQDTPEERCSVAAYVSARTMPEAKADDIVGPVTHEIFENNVVHLMWQEPKEPNGL
IVLYEVSYRRYGDEELHLCVSRKHFALERGCRLRGLSPGNYSVRIRATSLAGNGSWTEPTYFYVTDYLDVPSNIAKIIIG
PLIF

Stampare la sequenza su file.